File: pd_beaker_four_bits_of_memory_df.r
library (reshape)
library (ggplot2)
library (dplyr)


#THIS IS FOR BITS OF MEMORY OVER TIME

bits_of_memory_df <- read.csv("all_bits_df_static_comp_more_values.csv")
#bits_of_memory_df <- as.data.frame(beaker::get('all_bits_df'))

#add
colnames(bits_of_memory_df) <- c('Row_Label', 'B0', 'B1', 'B2', 'B3', 'B4','Condition', 'Generation')

bits_of_memory_df$Row_Label <- NULL
bits_of_memory_df$Generation <- as.numeric(as.character(bits_of_memory_df$Generation))
bits_of_memory_df$Condition <- as.factor(as.numeric(as.character(bits_of_memory_df$Condition)))
bits_of_memory_df$B0 <- as.numeric(as.character(bits_of_memory_df$B0))
bits_of_memory_df$B1 <- as.numeric(as.character(bits_of_memory_df$B1))
bits_of_memory_df$B2 <- as.numeric(as.character(bits_of_memory_df$B2))
bits_of_memory_df$B3 <- as.numeric(as.character(bits_of_memory_df$B3))
bits_of_memory_df$B4 <- as.numeric(as.character(bits_of_memory_df$B4))


#Use for testing
#Dummy df columns
#B0 <- c(17, 17, 17, 17, 17, 17, 17, 18)
#B1 <- c(17, 17, 17, 17, 17, 17, 17, 17)
#B2 <- c(18, 17, 17, 17, 17, 17, 17, 17)
#B3 <- c(17, 17, 17, 18, 17, 17, 17, 17)
#B4 <- c(17, 17, 17, 17, 17, 17, 18, 17)
#Generation <- c(500, 500, 500, 500, 500, 500, 500, 500)
#Condition <- c(3, 3, 3, 3, 7, 7, 7, 7)
#Make new dummy data frame to test pairwise wilcox
#dummy_df <- data.frame(B0, B1, B2, B3, B4, Condition, Generation)
#dummy_df$Row_Label <- NULL
#dummy_df$Condition <- as.factor(dummy_df$Condition)
#bits_of_memory_df <- rbind(bits_of_memory_df, dummy_df)

#bits_of_memory_df


weights = 0:4
bits_of_memory_df$Mean <- apply(bits_of_memory_df, 1, function(d) {weighted.mean(weights, d[1:5])})

#want a smaller df
#for given condition and generation what is the mean of  the means
#aggregate( Mean~Generation + Condition, bits_of_memory_df, function(x){c(group_mean = mean(x))})
summary_df <- group_by(bits_of_memory_df, Condition, Generation) 

summary_df <- summarize(summary_df, group_mean = mean(Mean), group_sd = sd(Mean) )
#standard deviation
#print(summary_df)
#bits_of_memory_df$Generation <- c(1:nrow(bits_of_memory_df))




ggplot(data = summary_df, aes(x = Generation, y = group_mean, fill = Condition, ymin = group_mean - group_sd, ymax = group_mean + group_sd)) + geom_line(aes(color = Condition)) + theme_minimal() + geom_ribbon(alpha=.3) + theme(panel.grid.major = element_blank(), panel.background = element_blank(), panel.grid.minor = element_blank()) + scale_y_continuous("Average Bits of Memory")

ggsave("Average_Bits_Memory_Overtime_Static_Comp.pdf")


ggplot(data = subset(summary_df, Condition %in% c(0, 0.01, 0.03, 0.05, 0.075)), aes(x = Generation, y = group_mean, fill = Condition, ymin = group_mean - group_sd, ymax = group_mean + group_sd)) + geom_line(aes(color = Condition)) + theme_minimal() + geom_ribbon(alpha=.3) + theme(panel.grid.major = element_blank(), panel.background = element_blank(), panel.grid.minor = element_blank()) + scale_y_continuous("Average Bits of Memory") + scale_fill_manual(values=c("#F5828C", "#8CF582", "#82F5EB", "#828BF5")) + scale_color_manual(values=c( "#F5828C", "#8CF582",  "#82F5EB", "#828BF5"))

ggsave("Average_Bits_Memory_Overtime_Restricted_Static_Comp.pdf")



#Makes assumptions about data
#Have to check these assumptions
#Each sample is an indep random sample
#Distribution follows normal distrib
#Equal width
#MAKE HISTOGRAM TO CHECK THESE ASSUMPTIONS
#hist(bits_of_memory_df$Mean)

#ANOVA TEST
#How different are things
#anova_bits <- aov(bits_of_memory_df$Mean ~ bits_of_memory_df$Condition) 
#summary(anova_bits)


#TUKEY TEST
#Off of ANOVA
#Things are different so which ones are different
#TukeyHSD(anova_bits)


#KRUSKAL-WALLIS TEST
#How different are things
#Similar to Anova
#Doesn't assume normality of shape of graph
kruskal_bits <- kruskal.test(Mean ~ Condition, data = subset(bits_of_memory_df, bits_of_memory_df$Generation == 499))
end_df <- subset(bits_of_memory_df, bits_of_memory_df$Generation == 499)
kruskal_bits

#Wilcox and Rank Sum Test
#Like TUKEY but doesn't assume normality
pairwise.wilcox.test(end_df$Mean, end_df$Condition, "bonferroni")

warnings()

File: pd_beaker_four_bits_of_memory_df.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from statsmodels.stats.multicomp import pairwise_tukeyhsd, MultiComparison

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning) #For those pesky deprecation warnings


bits_of_memory_df = pd.read_csv("all_bits_df_static_comp_more_values.csv")
bits_of_memory_df.replace([np.inf, -np.inf], np.nan, inplace=True)
bits_of_memory_df.columns = ['Row_Label', 'B0', 'B1', 'B2', 'B3', 'B4', 'Condition', 'Generation']
bits_of_memory_df.drop(['Row_Label'], axis=1, inplace=True)
bits_of_memory_df['Generation'] = pd.to_numeric(bits_of_memory_df['Generation'], errors='coerce')
bits_of_memory_df['Condition'] = pd.Categorical(pd.to_numeric(bits_of_memory_df['Condition'], errors='coerce'))
for col in ['B0', 'B1', 'B2', 'B3', 'B4']:
    bits_of_memory_df[col] = pd.to_numeric(bits_of_memory_df[col], errors='coerce')

weights = np.arange(5)
bits_of_memory_df['Mean'] = bits_of_memory_df.apply(lambda row: np.average(row[:5], weights=weights), axis=1)

summary_df = bits_of_memory_df.groupby(['Condition', 'Generation'], observed=True).agg(
    group_mean=('Mean', 'mean'),
    group_sd=('Mean', 'std')
).reset_index()

sns.lineplot(data=summary_df, x='Generation', y='group_mean', hue='Condition', palette='muted')
plt.fill_between(data=summary_df, x='Generation', y1=summary_df['group_mean'] - summary_df['group_sd'], y2=summary_df['group_mean'] + summary_df['group_sd'], alpha=0.3)
plt.title('Average Bits of Memory Over Time')
plt.ylabel('Average Bits of Memory')
plt.grid(False)
plt.savefig('Average_Bits_Memory_Overtime_Static_Comp.pdf')

filtered_df = summary_df[summary_df['Condition'].isin([0, 0.01, 0.03, 0.05, 0.075])]
sns.lineplot(data=filtered_df, x='Generation', y='group_mean', hue='Condition') #, palette=['#F5828C', '#8CF582', '#82F5EB', '#828BF5'] # Palette for when we have 4 samples
plt.fill_between(data=filtered_df, x='Generation', y1=filtered_df['group_mean'] - filtered_df['group_sd'], y2=filtered_df['group_mean'] + filtered_df['group_sd'], alpha=0.3)
plt.title('Average Bits of Memory Over Time (Restricted Conditions)')
plt.ylabel('Average Bits of Memory')
plt.grid(False)
plt.savefig('Average_Bits_Memory_Overtime_Restricted_Static_Comp.pdf')

plt.hist(bits_of_memory_df['Mean'], bins=30)
plt.title('Distribution of Mean Bits of Memory')
plt.xlabel('Mean Bits of Memory')
plt.ylabel('Frequency')
plt.savefig('Mean_Bits_Memory_Distribution_Check.png')

# anova_results = stats.f_oneway(*[group['Mean'].values for name, group in bits_of_memory_df.groupby('Condition', observed=True)])
# print(f'ANOVA test results: {anova_results}')

# tukey_results = pairwise_tukeyhsd(bits_of_memory_df['Mean'], bits_of_memory_df['Condition'])
# print(tukey_results)

# kruskal_data = bits_of_memory_df[bits_of_memory_df['Generation'] == 499]
# kruskal_results = stats.kruskal(*[group['Mean'].values for name, group in kruskal_data.groupby('Condition', observed=True)])
# print(f'Kruskal-Wallis test results: {kruskal_results}')

# comparison = MultiComparison(kruskal_data['Mean'], kruskal_data['Condition'])
# wilcox_results = comparison.allpairtest(stats.mannwhitneyu, method='bonferroni')
# print(wilcox_results[0])
